src/Main.hs

```diff
@@ -83,6 +94,31 @@ noEdges = AllEdges { normalEdges    = IMap.empty
                    , sourceEdges    = IMap.empty
                    }
 
+combs :: Int -> [([Int], [Int])] -> [[([Int], [Int])]]
+combs 0 _ = [[]]
+combs _ [] = []
+combs k (x:xs) = ((x :) <$> combs (k - 1) xs) ++ combs k xs
+
+toposort :: [(Int, [Int])] -> [Int]
+toposort xs
+  | (not . null) cycleDetect =
+    error $ "Dependency cycle detected for nodes " ++ show cycleDetect
+  | otherwise = foldl makePrecede [] dB
+  where
+    dB :: [([Int], [Int])]
+    dB = ((\(x, y) -> (x, y \\ x)) . (return *** id)) <$> xs
+    makePrecede :: [Int] -> ([Int], [Int]) -> [Int]
+    makePrecede ts ([x], xs) =
+      nub $
+      case elemIndex x ts of
+        Just i -> uncurry (++) $ first (++ xs) $ splitAt i ts
```

Comment:
Err... Also here, could you write in a more directly way without using `first` and `uncurry`